libantworld/agent.h

```diff
@@ -21,67 +35,279 @@ using namespace units::literals;
+                        stopMoving();
+                    }
+                    return true;
+                });
     }
 
-    template<class AngleUnit = degree_t>
-    Vector3<AngleUnit> getAttitude() const
+    template<typename LengthUnit = meter_t>
+    Vector3<LengthUnit> getPosition()
     {
-        return convertUnitArray<AngleUnit>(m_Attitude);
+        std::lock_guard<std::mutex> guard(m_PoseMutex);
+        updatePose();
+        return convertUnitArray<LengthUnit>(m_Pose.position());
+    }
+
+    template<typename AngleUnit = degree_t>
+    Array3<AngleUnit> getAttitude()
+    {
+        std::lock_guard<std::mutex> guard(m_PoseMutex);
+        updatePose();
+        return convertUnitArray<AngleUnit>(m_Pose.attitude());
+    }
+
+    template<typename LengthUnit = meter_t, typename AngleUnit = degree_t>
+    Pose3<LengthUnit, AngleUnit> getPose()
+    {
+        std::lock_guard<std::mutex> guard(m_PoseMutex);
+        updatePose();
+        return m_Pose;
+    }
+
+    GLFWwindow *getWindow() const
+    {
+        return m_Window;
+    }
+
+    radians_per_second_t getMaximumTurnSpeed()
+    {
+        return m_TurnSpeed;
     }
 
     void setPosition(meter_t x, meter_t y, meter_t z)
     {
-        m_Position[0] = x;
-        m_Position[1] = y;
-        m_Position[2] = z;
+        BOB_ASSERT(m_MoveMode == MoveMode::NotMoving);
+        m_Pose.position() = { x, y, z };
     }
 
     void setAttitude(degree_t yaw, degree_t pitch, degree_t roll)
     {
-        m_Attitude[0] = yaw;
-        m_Attitude[1] = pitch;
-        m_Attitude[2] = roll;
+        BOB_ASSERT(m_MoveMode == MoveMode::NotMoving);
+        m_Pose.attitude() = { yaw, pitch, roll };
     }
 
-    virtual bool readFrame(cv::Mat &frame) override
+    bool update()
     {
+        // If the agent is "moving", we need to calculate its current position
+        std::lock_guard<std::mutex> guard(m_PoseMutex);
+        updatePose();
+
+        // Render to m_Window
+        glfwMakeContextCurrent(m_Window);
+
         // Clear colour and depth buffer
         glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
 
         // Render first person
         const auto size = getOutputSize();
-        m_Renderer.renderPanoramicView(m_Position[0], m_Position[1], m_Position[2],
-                                       m_Attitude[0], m_Attitude[1], m_Attitude[2],
+        m_Renderer.renderPanoramicView(m_Pose.x(), m_Pose.y(), m_Pose.z(),
+                                       m_Pose.yaw(), m_Pose.pitch(), m_Pose.roll(),
                                        0, 0, size.width, size.height);
 
         // Swap front and back buffers
         glfwSwapBuffers(m_Window);
 
+        return true;
+    }
+
+    bool isOpen() const { return !glfwWindowShouldClose(m_Window); }
+
+    virtual bool readFrame(cv::Mat &frame) override
+    {
+        update();
+
         // Read frame
         return Video::OpenGL::readFrame(frame);
     }
 
+    virtual void stopMoving() override
```

Comment:
Is there any particular reason for the move mode stuff? It seems overly complicated in this instance (maybe it makes more sense for something else?) - why can't you move and turn at the same time and why do you have to not be moving in order to set position?

Reply:
I was using it for some simple simulations elsewhere but I'm not using any of the moving stuff in this branch atm.

You can only turn on the spot or move forwards at the moment, but we could change it to take an angular velocity and a forward velocity (I've got code for this with the robot positioner stuff). I didn't bother adding the code to update pose while the agent is moving, but it wouldn't be hard to do and maybe I should.

Reply:
Maybe let's just leave this as-is for the mo -- we can maybe tidy it up once the other simulation/robot positioner stuff is merged.